rsocket-core/src/commonMain/kotlin/io/rsocket/kotlin/internal/flow/LimitingFlowCollector.kt

```diff
@@ -14,18 +14,28 @@
  * limitations under the License.
  */
 
-package io.rsocket.kotlin.flow
+package io.rsocket.kotlin.internal.flow
 
+import io.rsocket.kotlin.payload.*
 import kotlinx.atomicfu.*
 import kotlinx.coroutines.*
+import kotlinx.coroutines.flow.*
 
-internal class LimitStrategy(override val initialRequest: Int) : RequestStrategy {
-    private val requests = atomic(initialRequest)
+internal abstract class LimitingFlowCollector(initial: Int) : FlowCollector<Payload> {
+    private val requests = atomic(initial)
     private val awaiter = atomic<CancellableContinuation<Unit>?>(null)
 
-    override suspend fun requestOnEmit(): Int {
+    abstract suspend fun emitValue(value: Payload)
+
+    fun updateRequests(n: Int) {
+        if (n <= 0) return
+        requests.getAndAdd(n)
+        awaiter.getAndSet(null)?.resumeSafely()
```

Comment:
Is that requester or the responder's side? 
What is the concurrency model there?
It looks a little weak to me, thus asking for clarification

Reply:
Both requester in responder. On requester side used for limiting request Flow from `requestChannel` , on responder side used for limiting response from `requestsStream` and `requestChannel`.
`emit` will be called sequantally, `updateRequests` will be called from different thread, when requestN frame received.
BTW, that implementation is based on [coroutines reactive integration](https://github.com/Kotlin/kotlinx.coroutines/blob/master/reactive/kotlinx-coroutines-reactive/src/ReactiveFlow.kt).